src/iterators/eachmer.jl

```diff
@@ -131,116 +131,120 @@ const kmerbits = (0xff, 0x00, 0x01, 0xff,
                   0x03, 0xff, 0xff, 0xff,
                   0xff, 0xff, 0xff, 0xff)
 
+@inline blank_ntuple(::Type{Kmer{A,K,N}}) where {A,K,N} = ntuple(x -> zero(UInt64), Val{N}())
+
 # Initializers for two-bit nucleic acid alphabets...
-@inline function Base.iterate(it::AbstractMerIterator{T,S}) where {T,S<:LongSequence{<:NucleicAcidAlphabet{2}}}
-    UT = encoded_data_type(T)
+@inline function Base.iterate(it::AbstractMerIterator{Kmer{A,K,N},S}) where {A,K,N,S<:LongSequence{<:NucleicAcidAlphabet{2}}}
     filled, i = 0, it.start
-    fwkmer = rvkmer = zero(UT)
+    fwkmer = rvkmer = blank_ntuple(Kmer{A,K,N})
 
     while i ≤ it.stop
         nt = reinterpret(UInt8, inbounds_getindex(it.seq, i))
-        @inbounds fbits = UT(kmerbits[nt + 1])
+        @inbounds fbits = UInt64(kmerbits[nt + 1])
         rbits = ~fbits & typeof(fbits)(0x03)
-        fwkmer = (fwkmer << 0x02 | fbits)
-        rvkmer = (rvkmer >> 0x02) | (UT(rbits) << unsigned(offset(T, 1)))
+        # fwkmer = (fwkmer << 0x02 | fbits)
+        fwkmer = leftshift_carry(fwkmer, 2, fbits)
+        #rvkmer = (rvkmer >> 0x02) | (UInt64(rbits) << unsigned(offset(T, 1)))
+        rvkmer = rightshift_carry(rvkmer, 2, (UInt64(rbits) << (62 - (64N - 2K))))
         filled += 1
-        if filled == ksize(T)
-            return MerIterResult(1, T(fwkmer), T(rvkmer)), (i, fwkmer, rvkmer)
+        if filled == K
+            return MerIterResult(1, Kmer{A,K,N}(fwkmer), Kmer{A,K,N}(rvkmer)), (i, fwkmer, rvkmer)
         end
         i += 1
     end
     return nothing
 end
 
-@inline function Base.iterate(it::EveryMerIterator{T,S}, state) where {T,S<:LongSequence{<:NucleicAcidAlphabet{2}}}
-    UT = encoded_data_type(T)
+@inline function Base.iterate(it::EveryMerIterator{Kmer{A,K,N},S}, state) where {A,K,N,S<:LongSequence{<:NucleicAcidAlphabet{2}}}
     i, fwkmer, rvkmer = state
     i += 1
-    
     if i > it.stop
         return nothing
     else
         nt = reinterpret(UInt8, inbounds_getindex(it.seq, i))
-        @inbounds fbits = UT(kmerbits[nt + 1])
+        @inbounds fbits = UInt64(kmerbits[nt + 1])
         rbits = ~fbits & typeof(fbits)(0x03)
-        fwkmer = (fwkmer << 0x02 | fbits)
-        rvkmer = (rvkmer >> 0x02) | (UT(rbits) << unsigned(offset(T, 1)))
-        pos = i - ksize(T) + 1
-        
-        return MerIterResult(pos, T(fwkmer), T(rvkmer)), (i, fwkmer, rvkmer)
+        #fwkmer = (fwkmer << 0x02 | fbits)
+        fwkmer = leftshift_carry(fwkmer, 2, fbits)
```

Comment:
Yes kmers only support 2-bit elements currently, as ambiguous nucleotides were never allowed, and so I just use 2, old code would also hardcode 2 or 0x02 in places too. Generalizing over all alphabets is one possible enhancement, but I don't want to spend time on it unless a person into amino acids say comes and says "yes please I really want this" - because whilst I'm doing assembly infrastructure it's just not a high priority for me.